src/Workspaces/Core/Portable/FlowAnalysis/SymbolUsageAnalysis/SymbolUsageAnalysis.AnalysisData.cs

```diff
@@ -0,0 +1,230 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis.SymbolUsageAnalysis
+{
+    internal static partial class SymbolUsageAnalysis
+    {
+        /// <summary>
+        /// Core analysis data to drive the operation <see cref="Walker"/>
+        /// for operation tree based analysis OR control flow graph based analysis.
+        /// </summary>
+        private abstract class AnalysisData : IDisposable
+        {
+            /// <summary>
+            /// Pooled <see cref="BasicBlockAnalysisData"/> allocated during analysis with the
+            /// current <see cref="AnalysisData"/> instance, which will be freed during <see cref="Dispose"/>.
+            /// </summary>
+            private readonly ArrayBuilder<BasicBlockAnalysisData> _allocatedBasicBlockAnalysisDatas;
+            
+            protected AnalysisData(
+                PooledDictionary<(ISymbol symbol, IOperation operation), bool> symbolWriteBuilder,
+                PooledHashSet<ISymbol> symbolsRead,
+                PooledHashSet<IMethodSymbol> lambdaOrLocalFunctionsBeingAnalyzed)
+            {
+                SymbolsWriteBuilder = symbolWriteBuilder;
+                SymbolsReadBuilder = symbolsRead;
+                LambdaOrLocalFunctionsBeingAnalyzed = lambdaOrLocalFunctionsBeingAnalyzed;
+
+                _allocatedBasicBlockAnalysisDatas = ArrayBuilder<BasicBlockAnalysisData>.GetInstance();
+                CurrentBlockAnalysisData = CreateBlockAnalysisData();
+            }
+
+            /// <summary>
+            /// Map from each (symbol, write) to a boolean indicating if the value assigned
+            /// at the write is read on some control flow path.
+            /// </summary>
```

Comment:
i think there's just enogh in this comment to understand what this means.  However, consdier examples in teh future to help clarify as this stuff can sometimes be vague.  i.e. you could show:

```c#
int a = 0;
a = 1;
a = 2;
Console.WriteLine(a);
```

Both the '0' and '1' writes are not ever read on any control flow paths, but the write of '2' is.

Reply:
Sure, done.

---
In reply to: [231668322](https://github.com/dotnet/roslyn/pull/31025#discussion_r231668322) [](ancestors = 231668322)